### ISSUE TITLE
OS Image Distribution from S3 to Customer vCenters

### ISSUE THREAD

**Issue by gawertm** [giantswarm/roadmap#2990]
We should implement a mechanism to distribute the images from the common place (AWS S3 now) to all vSphere and VCD environments automatically to be able to manage multiple customers/users easily

from: https://github.com/giantswarm/roadmap/issues/2757#issuecomment-1822303696

**Comment by gawertm** [comment]
@giantswarm/team-rocket how could that work? this would rather need to be a pull mechanism than a push because of network restrictions, right?

**Comment by glitchcrab** [comment]
Yes we would need to pull the image from the central location, most likely via an operator or maybe a cronjob in the cluster and then push it to the customer's catalog

**Comment by glitchcrab** [comment]
I want to bump this back up in priority - because all providers+customers are now using releases, the number of images we need to upload has increased. Previously we used to set the image template in the cluster-provider chart and it wasn't often updated, however the release-based charts see kubernetes and/or flatcar upgrades far more often and so we have to upload images more often (especially for test installations). Additionally, the number of locations we need to upload them too is increasing with more on-prem customers coming onboard. 

An operator shouldn't be _too_ complex to write - I suggest we provide it a whitelist of images to upload (so as to avoid us uploading every single image which is built). This has the added benefit that we don't need to have the operator watch the S3 bucket - it just reconciles the existing images in the vcenter/vcd catalog against the whitelist and then just add any which are missing.

**Comment by vxav** [comment]
That's awesome

**Comment by erkanerol** [comment]
I liked it too. <3

**Comment by vxav** [comment]
Hey, if Erkan approves it then you're good to do 🚀 

A few unrefined thoughts out of the blue (to be mindful of not wasting customer storage):

- Delete images older than the oldest WC/MC (with the pre-requesite of 1 MC : 1 vCenter). That way we automatically get rid of images used by skipped releases and older ones.
- With regards to ☝: Do we have a way to check `image name:Release` association?

**Comment by Gacko** [comment]
Could you please also consider Release PRs? New images need to be published to test environments so we can run Releases Test Suites. Therefore images of not yet released Releases need to be pushed to vSphere / Cloud Director, too.

Overall I'd prefer not being bound to Release CRs, because users can also create them for testing purposes and they even get created by Cluster / Releases Test Suites.

Is there no possibility to just sync what's in the S3 bucket? Maybe with a RegEx / name pattern so we only sync relevant ones.

**Comment by Gacko** [comment]
> With regards to ☝: Do we have a way to check image name:Release association?

The releases do not contain the image name, only components to concatenate it. The template is defined in the provider specific cluster charts.

**Comment by AverageMarcus** [comment]
> Could you please also consider Release PRs? New images need to be published to test environments so we can run Releases Test Suites. Therefore images of not yet released Releases need to be pushed to vSphere / Cloud Director, too.
> 
> Overall I'd prefer not being bound to Release CRs, because users can also create them for testing purposes and they even get created by Cluster / Releases Test Suites. So Release CRs might not be the best point of truth.
> 
> Is there no possibility to just sync what's in the S3 bucket? Maybe with a RegEx / name pattern so we only sync relevant ones.

The test suite creates new Release CRs on the MC so this would still pick it up, just might delay things a little. We can maybe also implement some status condition (`imageAvailable` or something) that we can wait on in the tests.

**Comment by anvddriesch** [comment]
> Delete images older than the oldest WC/MC (with the pre-requesite of 1 MC : 1 vCenter). That way we automatically get rid of images used by skipped releases and older ones.

You are referring to the deletion of images that are still technically part of a release which is present on the MC but aren't/ won't be in use, right?
I agree that we should not clog the whole catalog with unneeded images for sure. However, in case we are watching releases and syncing them to the images it would be simpler to clean up the older releases instead. Otherwise we could break cluster creation (or god forbid a downgrade 😢 ) for older releases available on the MC which is confusing.
Or well if we are already watching whether releases are in use or not we could reupload them again.
But that might result in a lot of back and forth and some wait times if the upload is slow. 🤔 

> The test suite creates new Release CRs on the MC so this would still pick it up, just might delay things a little. 
I admit that we would have the same wait/back and forth in this case.

> Is there no possibility to just sync what's in the S3 bucket? Maybe with a RegEx / name pattern so we only sync relevant ones.

We can probably do that as well but I see some downsides.
If we are not syncing with releases we will end up pulling all images for a specific provider. That results in a lot of unused images. 
We would need to come up with a system for cleaning up which... would likely involve checking releases anyway so that we don't have the above issue where a release is there and a cluster can not be created for it.
Ideally the available releases on an MC should be the source of truth for which images should be present.

If we do want to speed things up for tests and the like we could consider also pulling images that are newer than the newest release on test MCs.
In this case they would be there a bit quicker for their release to use them and then they'll be cleaned up later anyway. (when a newer release exists and their release is gone)
I would only do this for test MCs though.

**Comment by vxav** [comment]
A few things about VCD and vSphere for @anvddriesch (we can also discuss more later):

## vSphere

- The OVA is uploaded to vCenter as a VM and must be converted to template afterwards (there'll be a `govc` command for that if you use it).
- Templates/VMs are physically located (as files on shared storage) in a vSphere cluster (collection of ESXi hosts) and then stored organisationally in a "VM Folder". Sometimes the template must be deployed to multiple locations when there are several DCs. for instance with `sardine` there is one vCenter managing multiple datacenters (US, China, ....). In which case the template must be uploaded to both locations. e.g. 👇 So we'll need to be able to specify an array of objects to configure the destinations of this app.

<img width="613" alt="Image" src="https://github.com/user-attachments/assets/8071c6ed-ee6e-4f43-8f08-871464f14f6a" />

## VCD

- VCD has the concept of named catalogs which you probably know better. 
- Our catalog is always named `giantswarm`

**Comment by AverageMarcus** [comment]
> I would only do this for test MCs though.

Oh now that I think about it, having a flag in the operator that can be toggled on specifically for our stable-testing envs that causes it to copy all images would be quite useful. I suspect we're less bothered ourselves with the extra storage being used compared to our customers.

**Comment by Gacko** [comment]
Overall what I am concerned of is the need for a comparable fast and, more important, reliable internet connection. Some providers are charging you for down- and uploaded data these days and so I'd prefer a solution that's not down- and uploading comparable large files on-demand. Please also have edge computing on your mind when designing this. Just think of Manuel's home lab we're about to use for Proxmox. I heard the internet connection sometimes isn't that good there. 😉

**Comment by AverageMarcus** [comment]
One way of possibly handling it is to have the release logic only in the cleanup and implement a TTL for images copied over.  (Though I'm not 100% convinced of this)

Something like:
1. New image found in S3 bucket -> copy over to vsphere env
2. On some recurring basis (hourly, daily, whatever) check the age of all images that are in the vsphere env and for each...
	a. Check if the age of the image is older than a TTL (lets say 7 days)
	b. If it is, check if it matches any Releases currently on the MC and if not...
	c. Delete the image from vsphere

This means images will be available as soon as created for tests to use but won't stick around forever taking up space.

For extra safety (and scenarios where setting up new MCs / vsphere envs) still have the operator monitor Releases and ensure that images are available in the vsphere env and if not copy them over then.

I'm not sure we need anything complex like pattern-based matching of sync. Otherwise that'll get to the messy state of retagger which is totally overkill for what we're trying to achieve and our _current_ requirements. (No need to over-engineer for what-ifs)

Regarding the "reliable internet connection" comment - that should be a requirement of our platform anyway and I recall it being mentioned as such during the on-site discussions.

**Comment by anvddriesch** [comment]
I agree we should not over-engineer this. Let's start with a simple solution.
We have two basic scenarios.

1. A customers production MC - we should be concerned about space here and only supply images that are actually used. On the other hand we do not need to worry too much about speed in this case. I don't think people usually want to upgrade/create their cluster minutes after a new release is out. Obviously if it takes a really, really long time we should make adjustments.
2. Test MC - we are less concerned about uploading random stuff here since it's ourselves suffering, On the other hand we want test images to be available ASAP. For this pulling images early as we discussed and then cleaning up if too old/ not needed makes sense. If we have this mechanism it can be toggled for customer installations as well if it's important to have images instantly.

I personally think we can satisfy both of these with a controller based on watching release CRs. We will either way have to check what's in the bucket but that being the source of truth is an antipattern IMO.
As for slow/unavailable connection, we'll have this problem either way and it's fine as long as it reconciles eventually.

**Comment by anvddriesch** [comment]
I have to second this.
We have spend quite some time trying to implement our own set of functions to make the process faster and more verbose but it's been nothing but pain. 😄 
The importer version does the job.

**Comment by anvddriesch** [comment]
The first version `0.1.0` of `image-distribution-operator` has been released.
- it syncs images between the releases on the mc and vcenter.
- it maintains its own custom resources called `nodeImages` which show which releases they are part of and their current state (`pending`, `available` etc) in their status
- it imports images from a given s3 bucket to one or more locations in a given vcenter and convert to templates there.
- if an image is not available in s3 it will also show as `missing` in its status
- if there are no releases left which include a given image, the `nodeImage` cr will be deleted and templates will be cleaned up

Currently `ido` is part of `vsphere-app-collection` but none of the add-n (read hybrid) collections. So for now this is only running on test installations but we'll push it to hybrid as soon as we can confirm there are no issues.

There is some stuff left for future releases
- capvcd integration
- optionally including a location tag directly inside image names
- preloading images for tests (if tests aren't slowed down it  might be okay not to do that, lets see)
- including legacy names for templates (not sure if needed)

**Comment by vxav** [comment]
`ido` is broken in `wallaby` because the user password includes a `@` which breaks the [login url](https://github.com/giantswarm/image-distribution-operator/blob/main/pkg/vsphere/client.go#L66).

It seems [CAPV](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/550478a68a66df490602912bae680a257dfee3da/pkg/session/session.go#L116) authenticates slightly differently by working with sessions.

In the meantime, we can enforce a reduced set of special characters in vSphere passwords (we already hit this with them last year where it broke the CSI).

**Comment by anvddriesch** [comment]
I added two sub issues to address improvements on vsphere client and implementation of cloud director client!
